### PR TITLE
Fix #7703: make arrow function span the first line instead of the whole function body

### LINKED ISSUE #7703: VS/VS Code: Return type mismatch of an arrow function passed as an argument highlights the whole function body in red
Hopefully this is not a duplicate (couldn't find anything). Out of all the "less desirable" VS (and VS code as well) plugin behaviors I've found this to be the number one most disturbing:

**TypeScript Version:** 

1.8.7 (VS 2015 Community, Update 1)

**Screenshots**
clipboard02

And after fixing the return type (only changed `PromiseX.start<void>` to `PromiseX.start<any>`):
clipboard03

**Reduced test case**

```ts
function func(f: () => number) {
}

func(() => {
	let x = "HI"; // <-- Whole line highlighted in red
	return x;     // <-- Whole line highlighted in red
});
```
For comparison, when `function` is used instead, only the keyword "function" is highlighted:
```ts
function func(f: () => number) {
}

func(function() { // <-- only the "function" keyword is highlighted in red
	let x = "HI"; 
	return x;
});
```
**Suggestion**

No need to highlight the whole function body in this case. Highlighting only the arrow (i.e. `=>`) could also work (I tried to find a solution that would work with all variations of the arrow notation, e.g. including reduced notation like `val => val * 2`). This would be significantly less distracting and wouldn't mask other errors within the function scope.

On a more personal note, I would be super-thankful if this would be fixed! This happens very frequently when using callback or promise heavy code (I mean sometimes up to hundreds of times per day).

### PR DESCRIPTION
**Pull request by mhegazy**
Fixes #7703

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -364,6 +364,23 @@ namespace ts {
         return createTextSpanFromBounds(start, scanner.getTextPos());
     }
 
+    function getErrorSpanForArrowFunction(sourceFile: SourceFile, node: ArrowFunction): TextSpan {
+        const pos = skipTrivia(sourceFile.text, node.pos);
+
+        if (node.body && node.body.kind === SyntaxKind.Block) {
+            const {line: startLine } = getLineAndCharacterOfPosition(sourceFile, node.body.pos);
```
Comment (yuit) [inline src/compiler/utilities.ts]: add space before "line"? I think you may want to include this into linting

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -364,6 +364,23 @@ namespace ts {
...
+    function getErrorSpanForArrowFunction(sourceFile: SourceFile, node: ArrowFunction): TextSpan {
+        const pos = skipTrivia(sourceFile.text, node.pos);
+
+        if (node.body && node.body.kind === SyntaxKind.Block) {
+            const {line: startLine } = getLineAndCharacterOfPosition(sourceFile, node.body.pos);
+            const {line: endLine } = getLineAndCharacterOfPosition(sourceFile, node.body.end);
+            if (startLine < endLine) {
+                // The arrow function body spans multiple lines, 
+                // make the error span be the first line.
+                const endOfFirstLine = getLineStarts(sourceFile)[startLine + 1] - 1;
```
Comment (yuit) [inline src/compiler/utilities.ts]: I don't get why you need to do -1 here? shouldn't the etLineStarts(sourceFile)[startLine + 1] be enough?

**Comment by yuit** [comment]
Could you add the test case when arrow function span only one line?

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -179,6 +179,32 @@ namespace ts {
         return node.pos;
     }
 
+    export function getEndLinePosition(line: number, sourceFile: SourceFile): number {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Nit: consider swapping `sourceFile` and `line`

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -179,6 +179,32 @@ namespace ts {
...
+
+        const lineIndex = line;
+        if (lineIndex + 1 === lineStarts.length) {
+            // last line - return EOF
+            return sourceFile.text.length - 1;
+        }
+        else {
+            // current line start
+            const start = lineStarts[lineIndex];
+            // take the start position of the next line -1 = it should be some line break
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Space after `-`

File: src/compiler/utilities.ts [hunk 5]
```diff
@@ -179,6 +179,32 @@ namespace ts {
...
+            let pos = lineStarts[lineIndex + 1] - 1;
+            Debug.assert(isLineBreak(sourceFile.text.charCodeAt(pos)));
+            // walk backwards skipping line breaks, stop the the beginning of current line.
+            // i.e:
+            // <some text>
+            // $ <- end of line for this position should match the start position
+            while (start <= pos && isLineBreak(sourceFile.text.charCodeAt(pos))) {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Consider using a temp variable for `sourceFile.text` and use it elsewhere.

**Comment by DanielRosenwasser** [comment]
Changes like this make me really happy :clap: :clap: 

I left a few nits, but other than that :+1:

**Pull request merged**